core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -267,376 +302,264 @@ protected void doRun() throws Exception {
+                                    shardId, currentRoutingEntry, newShardRouting);
+                                indexService.removeShard(shardId.id(), "removing shard (recovery source node changed)");
+                            }
+                        }
+                    }
                 }
-                // clean the index
-                removeIndex(index, "removing index (no shards allocated)");
             }
         }
     }
 
-    private void applyIndexMetaData(ClusterChangedEvent event) {
-        if (!event.metaDataChanged()) {
+    private void createIndices(final ClusterChangedEvent event) {
+        // we only create indices for shards that are allocated
+        RoutingNode localRoutingNode = event.state().getRoutingNodes().node(event.state().nodes().getLocalNodeId());
+        if (localRoutingNode == null) {
             return;
         }
-        for (IndexMetaData indexMetaData : event.state().metaData()) {
-            if (!indicesService.hasIndex(indexMetaData.getIndex())) {
-                // we only create / update here
+        Set<Index> failedIndices = new HashSet<>(); // indices where creation failed (to make sure we only try creation once for each index)
+        for (ShardRouting shardRouting : localRoutingNode) {
+            if (failedShardsCache.containsKey(shardRouting.shardId())) {
+                // ignore
                 continue;
             }
-            // if the index meta data didn't change, no need check for refreshed settings
-            if (!event.indexMetaDataChanged(indexMetaData)) {
-                continue;
-            }
-            Index index = indexMetaData.getIndex();
-            IndexService indexService = indicesService.indexService(index);
-            if (indexService == null) {
-                // already deleted on us, ignore it
-                continue;
+            Index index = shardRouting.index();
+            if (failedIndices.contains(index)) {
+                // index creation for this index failed in an earlier iteration for another shard with same index
+                sendFailShard(shardRouting, "failed to create index (for shard with same index)", null);
+            } else if (indicesService.indexService(index) == null) {
+                final IndexMetaData indexMetaData = event.state().metaData().index(shardRouting.index());
+                logger.debug("[{}] creating index", index);
+                AllocatedIndex<? extends Shard> indexService = null;
+                try {
+                    indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, buildInIndexListener);
+                } catch (Throwable e) {
+                    sendFailShard(shardRouting, "failed to create index", e);
+                    failedIndices.add(index);
+                }
+                if (indexService != null) {
+                    updateIndexMapping(event, indexService, indexMetaData, failedIndices);
```

Comment:
future note - once refresh mapping is gone, we should inline this with index creation.